app/models/comment.rb

```diff
@@ -67,6 +72,25 @@ class Comment < ApplicationRecord
 
   after_commit :remove_from_elasticsearch, on: [:destroy]
 
+  # [@atsmith813] this is adapted from the `search_field` property in
+  # `config/elasticsearch/mappings/feed_content.json` and
+  # `app/serializers/search/comment_serializer.rb`
+  #
+  # highlighter settings are taken from
+  # Search::QueryBuildersFeedContent#add_highlight_fields
+  pg_search_scope :search_comments,
+                  against: %i[body_markdown],
+                  using: {
+                    tsearch: {
+                      prefix: true,
+                      highlight: {
+                        StartSel: "<mark>",
+                        StopSel: "</mark>",
+                        MaxFragments: 2
+                      }
+                    }
+                  }
+
```

Comment:
As mentioned [here](https://github.com/forem/forem/pull/13401/files#r613424752), polymorphism introduces some complexity here.

You may notice we're _only_ searching on the `body_text` field on `comments` and that with Elasticsearch we also search `user.username`, `user.name`, and `title`. 

That's because our manual joins due to polymorphism, the highlights, and our filters breaks the `associated_against` search in the `pg_search` gem. [Here](https://github.com/Casecommons/pg_search/issues/206#issuecomment-437614055) is a little more about that. So I'm only searching the fields on comments themselves (`body_text`). I'm not sure how common it is that people are searching for comments by username. I'd almost expect that case to be covered by a more direct user search.

Also, you may notice that [in the spike for this project (also pasted below for convenience)](https://github.com/forem/forem/pull/11985/files#diff-3baef9453c5d8b6c8962ce9ff9ec6b570b2e9cbad934e9f95b6d217ff1029b07R8-R34), we used `multisearchable` on this model to search. This had the advantage of searching through associations more easily and handling the additional attributes. It looked like this:

```ruby
  include PgSearch::Model
  # By treating PgSearch more akin to ES we need to store data during the index building step instead of joining it
  # `associated_against` also seems not to be effective during multisearch
  multisearchable against: %i[body_markdown pg_search_commentable_title pg_search_user_name 
  pg_search_user_username],
                  additional_attributes: lambda { |comment|
                    {
                      hotness_score: comment.score,
                      published: comment.commentable&.published,
                      published_at: comment.created_at,
                      public_reactions_count: comment.public_reactions_count
                    }
                  },
                  unless: ->(comment) { comment.deleted? || comment.hidden_by_commentable_user? || !comment.commentable&.published }
                  # order_within_rank: "score DESC, hotness_score DESC, comments_count DESC"

  def pg_search_commentable_title
    commentable&.title
  end

  def pg_search_user_name
    user.name
  end

  def pg_search_user_username
    user.username
  end
```

I chose _not_ to take that same approach because, [as @rhymes found](https://github.com/forem/forem/pull/13022), the serialization of the results seems to be the most common performance bottleneck. So while the approach taken in the spike makes the search easier, we would face issues in performantly serializing the results due to all the polymorphic associations/joins that would have to then be made afterward. I'm open to ideas on how to make this better, we'll this challenge again in other searches.

Reply:
Great idea! I'll follow up with Product to see what they think.